src/hdf5plugin/_filters.py

```diff
@@ -482,6 +486,112 @@ def __init__(self,
+            data=numpy.random.random(1000).reshape(100, 10),
+            **hdf5plugin.Sperr(rate=10))
+
+    - **Fixed peak signal-to-noise ratio (PSNR)**:
+      The quality argument provides the target PSNR (range: 0.0 < quality)
+
+      .. code-block:: python
+
+        f.create_dataset(
+            'sperr_fixed_peak_signal-to-noise_ratio',
+            data=numpy.random.random(1000).reshape(100, 10),
+            **hdf5plugin.Sperr(peak_signal_to_noise_ratio=1e-6))
+
+    - **Fixed point-wise error (PWE)**:
+      The quality argument provides the PWE tolerance (range: 0.0 < quality)
+
+      .. code-block:: python
+
+        f.create_dataset(
+            'sperr_fixed_point-wise_error',
+            data=numpy.random.random(1000).reshape(100, 10),
+            **hdf5plugin.Sperr(point_wise_error=1e-4))
+
+    For more details, see `H5Z-SPERR <https://github.com/NCAR/H5Z-SPERR>`_.
+    """
+    filter_name = "sperr"
+    filter_id = SPERR_ID
+
+    _FRACTIONAL_BITS = 16
+    _INTEGER_BITS = 12
+
+    def __init__(
+            self,
+            rate: float | None = None,
+            peak_signal_to_noise_ratio: float | None = None,
+            point_wise_error: float | None = None):
+        if (rate, peak_signal_to_noise_ratio, point_wise_error).count(None) < 2:
+            raise TypeError("hdf5plugin.Sperr() takes at most one not None argument")
+
+        if peak_signal_to_noise_ratio is not None:
+            assert peak_signal_to_noise_ratio > 0
+            mode = 2
+            quality = peak_signal_to_noise_ratio
+        elif point_wise_error is not None:
+            assert point_wise_error > 0
+            mode = 3
+            quality = point_wise_error
+        else:
+            assert rate is None or 0 < rate < 64
+            mode = 1
+            quality = 16 if rate is None else rate
+
+        self.filter_options = self.__pack_options(mode, quality)
+
+    @classmethod
+    def __pack_options(cls, mode: int, quality: float) -> tuple[int]:
+        assert mode in (1, 2, 3)
+        assert quality > 0
+
+        if mode in (1, 2):
+            ret = int(round(quality * (1 << cls._FRACTIONAL_BITS)))
+        else:  # mode == 3
+            quality_log = math.log2(quality)
+            if quality_log < 0:
+                ret = int(math.ceil(abs(quality_log) * (1 << cls._FRACTIONAL_BITS)))
+                # Store negative sign
+                ret |= 1 << (cls._INTEGER_BITS + cls._FRACTIONAL_BITS -1)
+            else:
+                ret = int(math.floor(quality_log * (1 << cls._FRACTIONAL_BITS)))
+
+        # encode mode in the top 4 bits
+        if mode == 1:
+            mask = 1 << (cls._INTEGER_BITS + cls._FRACTIONAL_BITS)
+        elif mode == 2:
+            mask = 1 << (cls._INTEGER_BITS + cls._FRACTIONAL_BITS)
+        else:  # mode == 3
+            mask = 1 << (cls._INTEGER_BITS + cls._FRACTIONAL_BITS)
+            mask |= 1 << (cls._INTEGER_BITS + cls._FRACTIONAL_BITS)
+        ret |= mask
```

Comment:
This needs to be double-checked

Reply:
Sorry that I changed added one more data field to be encoded in this function (you'll need to do a git pull to see the latest changes), so the function signature should look like:
```python
def __pack_options(cls, mode: int, quality: float, swap int) -> tuple[int]:
```
I have a few other comments:
- Line 586 should be: `mask = 1 << (cls._INTEGER_BITS + cls._FRACTIONAL_BITS + 1)`
- Line 589 should be: `mask |= 1 << (cls._INTEGER_BITS + cls._FRACTIONAL_BITS + 1)`
- Because of the added data field (`swap`) to be encoded, these lines are also needed:
```python
# encode the swap flag at bit 31
if swap != 0:
    ret |= 1 << (cls._INTEGER_BITS + cls._FRACTIONAL_BITS + 3)
```

Reply:
Thanks for the review! A second check was indeed needed!

Reply:
This was making the tests fails (mode was always 1)